MetaMask/metamask-extension
# Add ud resolution

## Explanation

Implements Unstoppable Domains Resolution Library
- Resolves UD on send and swap flow
- Resolves UD IPFS websites in address bar

## More Information

Responds to issue #12990

## Manual Testing Steps

[External MetaMask + UD Resolution Testing.pdf](https://github.com/MetaMask/metamask-extension/files/9559866/External.MetaMask.%2B.UD.Resolution.Testing.pdf)

### Testing Address Resolution

1. From the main screen press ‘Send’
2. Type Unstoppable Domain into search bar
3. Check to see if it resolves
4. It should resolve correctly for all UD TLDs
- .crypto
- .blockchain
- .bitcoin
- .zil
- .coin
- .nft 
- .wallet
- .888
- .dao
- .x

It should resolve for different currencies and tokens like MATIC & USDT

It should throw an error if
- The UD is not registered / is not valid OR if there is no address associated with the selected currency 
- The UD does not have an address associate with the selected currency / token
- After selecting a UD and the user swaps to a token that the UD doesn’t have an address for 

It should throw a warning if
- The UD contains a confusing character

### Testing IPFS Website Resolution
1. Simply type the UD into your browser’s address bar with a ‘/’ at the end 
2. On Chrome, Brave or other chrome based browsers it should
- Resolve all UD TLDs EXCEPT .888
- Redirect to IPFS website if a site exists
- Redirect to UD search page if IPFS website does not exist
3. On Firefox it should
- Resolve all UD TLDs
- Redirect to UD search page if IPFS website does not exist

## Pre-Merge Checklist


### + If there are functional changes:
